### ISSUE TITLE
RNaD implementation different than paper.

### ISSUE THREAD

**Issue by baskuit** [deepmind/open_spiel#961]
I might have to edit this later to better illustrate what I mean.

![image](https://user-images.githubusercontent.com/77259253/201453206-cbb5a0f9-d46a-4cb2-a55e-399c00d67651.png)

This is the formula for the estimate of the state value and Q-value used in DeepNash

![image](https://user-images.githubusercontent.com/77259253/201453274-931d2b39-53ef-4790-92b4-3eead564bd49.png)
And here are these estimates used in an update calculation.

Briefly, we can see in the picture the update is basically the logit of an action times the q-value of that action.
However the code used in the policy update is very different

        adv_pi = q_vr - torch.sum(pi * q_vr, dim=-1, keepdim=True)

        adv_pi = is_c * adv_pi  # importance sampling correction

        adv_pi = torch.clip(adv_pi, min=-clip, max=clip)

        adv_pi = adv_pi.detach()
        logits = logit_pi - torch.mean(logit_pi * legal_actions, dim=-1, keepdim=True)

        threshold_center = torch.zeros_like(logits)

        nerd_loss = torch.sum(

            legal_actions * apply_force_with_threshold(logits, adv_pi, threshold, threshold_center), axis=-1

        )

        nerd_loss = -renormalize(nerd_loss, valid * (player_ids == k))

        loss_pi_list.append(nerd_loss)

Here we can see the gradient is in the direction of the logit times the *advantage*. Additionally, we are also translating the logits so they have mean 0. This was not mentioned in the paper.
More importantly I see this `eta_reg_entropy` term in the code, which seems to be the reward shaping term:
`eta_reg_entropy = -eta * torch.sum(merged_policy * merged_log_policy, dim=-1) * torch.squeeze(player_others, dim=-1)`

![image](https://user-images.githubusercontent.com/77259253/201453484-7fb3c621-6aeb-48f6-9c9a-5802e8fef757.png)
However this is not quite the same as the update in the paper. In the code, we seem to have this extra `merged_policy` (pi) term in the multiplication of eta.

I've been putting this question off for a while, so sorry if this seems hasty

**Comment by lanctot** [comment]
Tagging the original contributors: @perolat @bartdevylder 

Thanks @baskuit, there's one thing I'm curious about first: you're quoting some code, but that code uses torch, whereas our RNAD implementation is based on Jax-- so the code you're quoting is not our code?

**Comment by baskuit** [comment]
Yeah sorry I quoted my pytorch adaptation by mistake. I'll edit in a bit. I just cant reconcile the _loop_v_trace function with the formulas in the paper.

**Comment by perolat** [comment]
Hi @baskuit,

I just updated the text in our internal version and it will be updated publicly by the end of the year I hope (sorry we forgot this average). The second part of the loss looks like this now:
![688570d875bb71ecfc432c034bf62fd1](https://user-images.githubusercontent.com/61795055/201694338-7fa18247-70c7-46d5-8792-00602d46ca2f.png)
We did our best to make sure we transcribed the code to comprehensible pseudocode and equations in the paper but we definitely missed this average step.

I'll take a bit more time to check your second point.

Hope it helps and thanks a lot for taking the time!
Julien
